astropy/timeseries/tests/test_downsample.py

```diff
@@ -33,38 +36,125 @@ def test_timeseries_invalid():
-    assert exc.value.args[0] == ("time_bin_size should be a astropy.unit quantity")
+
+def test_time_bin_invalid():
+
+    # Make sure to raise the right exception when time_bin_* is passed incorrectly.
+
+    with pytest.raises(TypeError, match=r"'time_bin_size' should be a Quantity or a TimeDelta"):
+        aggregate_downsample(TimeSeries(), time_bin_size=1)
+
+
+def test_binning_arg_invalid():
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+    with pytest.raises(TypeError, match=r"With single 'time_bin_start' either 'n_bins', "
+                                         "'time_bin_size' or time_bin_end' must be provided"):
+        aggregate_downsample(ts)
+
+
+def test_time_bin_conversion():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # Make sure time_bin_start and time_bin_end are properly converted to Time
+    down_start = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31'],
+                                      time_bin_size=[1]*u.s)
+    assert_equal(down_start.time_bin_start.isot, ['2016-03-22T12:30:31.000'])
+
+    down_end = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31', '2016-03-22T12:30:33'],
+                                    time_bin_end='2016-03-22T12:30:34')
+    assert_equal(down_end.time_bin_end.isot, ['2016-03-22T12:30:33.000', '2016-03-22T12:30:34.000'])
+
+
+def test_time_bin_end_auto():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # Interpret `time_bin_end` as the end of timeseries when `time_bin_start` is
+    # an array and `time_bin_size` is not provided
+    down_auto_end = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31', '2016-03-22T12:30:33'])
+    assert_equal(down_auto_end.time_bin_end.isot, ['2016-03-22T12:30:33.000', '2016-03-22T12:30:35.000'])
+
+
+def test_time_bin_start_array():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # When `time_bin_end` is an array and `time_bin_start` is not provided, `time_bin_start` is converted
+    # to an array with its first element set to the start of the timeseries and rest populated using
+    #`time_bin_end`. This case is separately tested since `BinnedTimeSeries` allows `time_bin_end` to
+    # be an array only if `time_bin_start` is an array.
+    down_start_array = aggregate_downsample(ts, time_bin_end=['2016-03-22T12:30:33', '2016-03-22T12:30:35'])
+    assert_equal(down_start_array.time_bin_start.isot, ['2016-03-22T12:30:31.000', '2016-03-22T12:30:33.000'])
+
+
+def test_nbins():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # n_bins should default to the number needed to fit all the original points
+    down_nbins = aggregate_downsample(ts, n_bins=2)
+    assert_equal(down_nbins.time_bin_start.isot, ['2016-03-22T12:30:31.000', '2016-03-22T12:30:33.000'])
 
 
+@pytest.mark.skipif(np.finfo(float).dtype == np.float32,
```

Comment:
This is surprisingly not getting skipped in the 32 bit run.

Reply:
> This is surprisingly not getting skipped in the 32 bit run. Maybe that has been the issue all this while.

Strange. Yes, would be interesting if the bin sizes would work with `sys.maxsize` as well, but I am leaving that up to you to decide; I’ll approve it either way.

Reply:
I'm fine either way. Might be easier to merge this and figure out how to change tests later for 32 bit.

Reply:
> I'm fine either way. Might be easier to merge this and figure out how to change tests later for 32 bit.

If
```python
    if sys.maxsize <= 2**32:
        time_bin_incr = (1 - 1e-6) * u.s
```
worked right away, it probably would be worth having a functional test, but if not, I'd just revert to the current HEAD and get that merged.

Reply:
Great I'll add that in!

Reply:
It works better than before: The failing test shows a difference in only the first element. But it still fails.